opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/ApplicationStatusRestService.java

```diff
@@ -0,0 +1,281 @@
+            }
+        }
+
+        return uptimeMillis / totalTimeMillis;
+    }
+
+    public double calculateApplicationPercentageUptime(final List<DowntimeInterval> intervals, final long start, final long end) {
+        double totalTimeMillis = end - start;
+        double uptimeMillis = totalTimeMillis;
+        for(final DowntimeInterval interval : intervals) {
+            uptimeMillis -= (Math.min(interval.end, end)-Math.max(interval.start, start));
+        }
+        return uptimeMillis / totalTimeMillis;
+    }
+
+
+    private ApplicationStatus getApplicationStatus(final OnmsApplication onmsApplication, final Collection<OnmsLocationSpecificStatus> statusChanges, final long start, final long end) {
+        final Map<OnmsMonitoringLocation, Map<OnmsMonitoredService, List<OnmsLocationSpecificStatus>>> m = new HashMap<>();
+
+        for(final OnmsLocationSpecificStatus onmsLocationSpecificStatus : statusChanges) {
+            m.computeIfAbsent(onmsLocationSpecificStatus.getLocation(), k -> new HashMap<>())
+                    .computeIfAbsent(onmsLocationSpecificStatus.getMonitoredService(), k -> new ArrayList<>())
+                    .add(onmsLocationSpecificStatus);
+        }
+
+        final ApplicationStatus applicationStatus = new ApplicationStatus();
+        applicationStatus.setStart(start);
+        applicationStatus.setEnd(end);
+        applicationStatus.setApplicationId(onmsApplication.getId());
+
+        for (final OnmsMonitoringLocation onmsMonitoringLocation : m.keySet()) {
+            final Location location = new Location();
+            location.setName(onmsMonitoringLocation.getLocationName());
+
+            final List<DowntimeInterval> downtimeIntervals = Lists.newArrayList();
+            
+            for (final OnmsMonitoredService onmsMonitoredService : m.get(onmsMonitoringLocation).keySet()) {
+                calculateServicePercentageUptime(m.get(onmsMonitoringLocation).get(onmsMonitoredService), downtimeIntervals, start, end);
+            }
+
+            final List<DowntimeInterval> mergedDowntimeIntervals = mergeDowntimeIntervals(downtimeIntervals);
+
+            location.setAggregatedStatus(100.0 * calculateApplicationPercentageUptime(mergedDowntimeIntervals, start, end));
+            applicationStatus.getLocations().add(location);
+        }
+        return applicationStatus;
+    }
+
+    private ApplicationServiceStatus getApplicationServiceStatus(final OnmsApplication onmsApplication, final Integer monitoredServiceId, final Collection<OnmsLocationSpecificStatus> statusChanges, final long start, final long end) {
+        final Map<OnmsMonitoringLocation, List<OnmsLocationSpecificStatus>> m = new HashMap<>();
+        final OnmsMonitoredService onmsMonitoredService = monitoredServiceDao.get(monitoredServiceId);
+
+        for(final OnmsLocationSpecificStatus onmsLocationSpecificStatus : statusChanges) {
```

Comment:
Is this collection required? Why not just calculate the values and fill the `ApplicationServiceStatus`?